### ISSUE TITLE
Remove unsafe Handle::new API

### ISSUE THREAD

**Issue by jdm** [servo/rust-mozjs#519]
https://github.com/servo/rust-mozjs/blob/5a50c377a4e1f55abc1286b74100a4cc35adc789/src/rust.rs#L771-L773

**Comment by jdm** [comment]
Specifically, if you have something like:
```rust
rooted!(in(cx) let foo = JS_Something());
let foo_handle = Handle::new(&*foo);
```
It's not clear to me if the result is meaningful. We should check what code uses this API in Servo.

**Comment by CYBAI** [comment]
I did a quick search and looks like we only use it in `components/script/webdriver_handlers.rs` 👀

<img width="497" alt="image" src="https://user-images.githubusercontent.com/6782666/88198942-39272100-cc7f-11ea-866a-19ff1f761d60.png">

**Comment by CYBAI** [comment]
Hmm, looks like this one can be rewritten as `.handle()`? 🤔

https://github.com/servo/servo/blob/6a9aac3e654d4498c8c2605ebe8a24b609770e21/components/script/webdriver_handlers.rs#L251-L259

**Comment by jdm** [comment]
I agree. And I think one clear source of unsafety from this API right now is:
```
let val = ObjectValue(some_rooted_object);
some_method_that_expects_a_handle_to_a_rooted_value(Handle::new(&val));
```
We should definitely remove it.

**Comment by supra08** [comment]
Hi @jdm I am looking towards contributing to this project. Would like to take this up as the first issue.

**Comment by jdm** [comment]
@supra08 That would be great!

**Comment by CYBAI** [comment]
I just saw @supra08 asked a question to @jdm about `what "remove" means` here. I wanted to answer the question that it actually means just remove the `pub fn new` function for `Handle` type.

However, I just noticed

https://github.com/servo/rust-mozjs/blob/5a50c377a4e1f55abc1286b74100a4cc35adc789/src/rust.rs#L771-L781

`Handle::new` will be called by `Handle::from_marked_location` and `Handle::from_marked_location`will be called by `Handle::from_raw`.

Then, with searching in Servo, I found we use `Handle::from_raw` in Servo more than `Handle::new`.

<img width="491" alt="image" src="https://user-images.githubusercontent.com/6782666/91183001-e4674200-e725-11ea-878b-cd9ffc179243.png">

If `remove` means "do not expose the new function", then it might be fine. but if `remove` means "remove the new function", then we might want to take care about how we want to handle `Handle::from_raw` without `Handle::new`?

**Comment by sudarshan-reddy** [comment]
Ah noted. I noticed the usage of from_raw as well and hence asked the question. Also noticed my user name seemed similar to @supra08 . Haha. I wanted to pick it up since it seemed inactive and I had free time.

Will wait on what you guys deem is the right way to remove it and submit a PR.

**Comment by jdm** [comment]
We still need to be able to create Handle values from from_marked_location and from_raw. It's really the fact that new is unsafe that makes it dangerous for outside code. Let's make the `new` method private, since it's important for `RootedGuard::handle()`.

**Comment by CYBAI** [comment]
> Also noticed my user name seemed similar to @supra08 . Haha. I wanted to pick it up since it seemed inactive and I had free time.

oops, sorry, I didn't notice that's not your account 😂

**Comment by sudarshan-reddy** [comment]
Noting for posterity that the usage of New under pub handle() implementations (here for example https://github.com/servo/rust-mozjs/blob/master)/src/rust.rs#L507 seems safe like @CYBAI indicated earlier as well.

**Comment by sudarshan-reddy** [comment]
https://github.com/servo/rust-mozjs/pull/523 addresses this. Build and tests passed locally. I felt it would be gratuitous to add a test to just marking a function private but tried to give a reasonably detailed commit message. 

Do let me know if you guys think this works. I'm happy to take any feedback and pointers you have as well.

**Comment by CYBAI** [comment]
This is fixed by #523

**Comment by sudarshan-reddy** [comment]
@jdm , @CYBAI : Is the next step to re-write components/script/webdriver_handlers.rs?

**Comment by CYBAI** [comment]
> @jdm , @CYBAI : Is the next step to re-write components/script/webdriver_handlers.rs?

Yes! As mentioned in https://github.com/servo/rust-mozjs/issues/519#issuecomment-662541274, also another thing you will need to do is updating the `rust-mozjs` revision to support "private Handle::new".

**Comment by sudarshan-reddy** [comment]
Sure thing @CYBAI . You mean the `rust-mozjs` rev that's used by `servo` correct?

**Comment by CYBAI** [comment]
Yes

**Comment by sudarshan-reddy** [comment]
Would I need to make a new issue on `servo` or use this one?

**Comment by jdm** [comment]
Continuing to use this one sounds fine.

**Comment by sudarshan-reddy** [comment]
So I updated servo and and changed the call to a `.handle()`. However, I ran into a bunch of compilation errors all which seem related to bindings. 

```
CompileOptionsWrapper::new(*cx, url_serialized.as_ptr(), handler.line as u32)
    |                                             ^^^^^^^^^^^^^^^^^^^^^^^ expected `&str`, found *-ptr
```

I noticed there's a moz_sys update as well (mozjs_sys v0.68.2). Will spend some time tomorrow to troubleshoot this and send in a PR.

**Comment by sudarshan-reddy** [comment]
@jdm , @CYBAI I discovered the root cause for the second error. (I think). 

[This change](https://github.com/servo/rust-mozjs/commit/831e7a390d2747bb18bef2dbc00e98ec583a2744#diff-35945a21f6ff748785df78b85fae7072R244) adds 2 more fields to NewProxyObject (classp and singleton) but the code generator [here](https://github.com/servo/servo/blob/master/components/script/dom/bindings/codegen/CodegenRust.py#L2776) does not account for this.

**Comment by sudarshan-reddy** [comment]
The fix for the 1st error seems easy (ref: https://github.com/servo/rust-mozjs/commit/38b7857704436a1c8ee4b5b96e62a16ecbb5293d). I just have to rewrite the `as_ptr()` to `&str` like shown in this PR.

**Comment by CYBAI** [comment]
> My question is, should it be part of this issue?

I think maybe it's worth filing a new issue in Servo because you will need to upgrade mozjs that you will need to fix both the `CompileOptionsWrapper` change and the `NewProxyObject` change in the mean while.

**Comment by CYBAI** [comment]
@sudarshan-reddy thanks a lot for fixing this and shipping it into Servo! I think this issue can be closed now :)

**Comment by sudarshan-reddy** [comment]
Thanks for all the help @CYBAI  and @jdm . I'll try picking up a new one tomorrow.
